compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -0,0 +1,802 @@
+    return op->emitError() << "invalid last stride, should be 1";
+  }
+  for (int i = 0; i < offsets.size(); ++i)
+    staticOffsets[staticOffsets.size() - offsets.size() + i] =
+        getConstantIntValue(offsets[i]).value();
+  for (int i = 0; i < sizes.size(); ++i)
+    staticSizes[staticSizes.size() - sizes.size() + i] =
+        getConstantIntValue(sizes[i]).value();
+  for (int i = 0; i < strides.size() - 1; ++i)
+    staticStrides[staticStrides.size() - (strides.size() - 1) + i] =
+        getConstantIntValue(strides[i]).value();
+  return success();
+}
+
+/// Convert the `amdaie.npu.dma_cpy_nd` operation to `aiex.npu.dma_memcpy_nd`.
+LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
+                                 AMDAIE::NpuDmaCpyNdOp dmaOp,
+                                 SmallVector<Operation *> &toBeErased,
+                                 IRMapping &mapper, IRMapping &bindingsMapper) {
+  rewriter.setInsertionPoint(dmaOp);
+  // Convert bidirectional dma copy nd op into two halves.
+  if (dmaOp.hasSourceAddressing()) {
+    SmallVector<Value> empty;
+    SmallVector<int64_t, 4> staticOffsets(4, 1);
+    SmallVector<int64_t, 4> staticSizes(4, 1);
+    SmallVector<int64_t, 3> staticStrides(3, 1);
+    if (failed(getStaticDims(dmaOp, dmaOp.getSourceMixedOffsets(),
+                             dmaOp.getSourceMixedSizes(),
+                             dmaOp.getSourceMixedStrides(), staticOffsets,
+                             staticSizes, staticStrides))) {
+      return failure();
+    }
+
+    AMDAIE::CircularDmaCpyNdOp dmaCpyNd = dmaOp.getDmaCpyNdOp();
+    auto memref =
+        bindingsMapper.lookup(dmaCpyNd.getSourceObjectFifo().getMemref());
+    auto symbol = dyn_cast<xilinx::AIE::ObjectFifoCreateOp>(
+                      mapper.lookup(dmaCpyNd.getOperation()))
+                      .getName();
+    // TODO(jornt): use bd_id != 0
+    bool issueToken = dmaOp.hasDmaWaitOpUser();
+    rewriter.create<AIEX::NpuDmaMemcpyNdOp>(
+        rewriter.getUnknownLoc(), SmallVector<Type, 1>{}, 0, 0, memref, empty,
+        empty, empty, staticOffsets, staticSizes, staticStrides, symbol, 0,
+        issueToken);
+  }
+  if (dmaOp.hasTargetAddressing()) {
+    SmallVector<Value> empty;
+    SmallVector<int64_t, 4> staticOffsets(4, 1);
+    SmallVector<int64_t, 4> staticSizes(4, 1);
+    SmallVector<int64_t, 3> staticStrides(3, 1);
+    if (failed(getStaticDims(dmaOp, dmaOp.getTargetMixedOffsets(),
+                             dmaOp.getTargetMixedSizes(),
+                             dmaOp.getTargetMixedStrides(), staticOffsets,
+                             staticSizes, staticStrides))) {
+      return failure();
+    }
+    auto dmaCpyNd = dmaOp.getDmaCpyNdOp();
+    auto memref =
+        bindingsMapper.lookup(dmaCpyNd.getTargetObjectFifo().getMemref());
+    auto symbol = dyn_cast<xilinx::AIE::ObjectFifoCreateOp>(
+                      mapper.lookup(dmaCpyNd.getOperation()))
+                      .getName();
```

Comment:
Same comment as above and also for similar cases below.

Reply:
Done